File: dart/lib/src/twitter/change_twitter_registry_data.dart
import 'dart:typed_data';

import '../constants/addresses.dart';
import '../instructions/instruction_types.dart';
import '../instructions/update_name_registry_instruction.dart';
import '../utils/get_domain_key_sync.dart';

/// Change Twitter registry data instruction
///
/// This function mirrors js/src/twitter/changeTwitterRegistryData.ts
///
/// Overwrites the data that is written in the user facing registry.
/// Must be signed by the verified pubkey.
///
/// [twitterHandle] - The Twitter handle to update data for
/// [verifiedPubkey] - The verified public key that can update the registry
/// [offset] - The offset at which to write the input data into the NameRegistryData
/// [inputData] - The new data to write
///
/// Returns a list of transaction instructions
Future<List<TransactionInstruction>> changeTwitterRegistryData(
  String twitterHandle,
  String verifiedPubkey,
  int offset,
  Uint8List inputData,
) async {
  // Hash the Twitter handle
  final hashedTwitterHandle = getHashedNameSync(twitterHandle);

  // Get the Twitter handle registry key using the name account key derivation
  final twitterHandleRegistryKey = await getNameAccountKeySync(
    hashedTwitterHandle,
    nameParent: twitterRootParentRegistryAddress,
  );

  // Create the update instruction
  final updateInstr = UpdateNameRegistryInstruction(
    offset: offset,
    inputData: inputData,
  );

  final instruction = updateInstr.getInstruction(
    programAddress: nameProgramAddress,
    domainAddress: twitterHandleRegistryKey,
    signer: verifiedPubkey,
  );

  return [instruction];
}

File: js/src/twitter/changeTwitterRegistryData.ts
import { Buffer } from "buffer";
import { PublicKey, TransactionInstruction } from "@solana/web3.js";
import {
  NAME_PROGRAM_ID,
  TWITTER_ROOT_PARENT_REGISTRY_KEY,
} from "../constants";
import { updateInstruction } from "../instructions/updateInstruction";
import { getHashedNameSync } from "../utils/getHashedNameSync";
import { getNameAccountKeySync } from "../utils/getNameAccountKeySync";
import { Numberu32 } from "../int";

/**
 * Builds an instruction that overwrites bytes in a verified Twitter registry.
 *
 * The verified public key must sign the resulting instruction.
 *
 * @param twitterHandle Verified Twitter handle whose registry is updated.
 * @param verifiedPubkey Signer that owns the verified registry.
 * @param offset Byte offset at which to write the data.
 * @param input_data Bytes to write into the registry data.
 * @returns The instruction that updates the registry data.
 *
 * @example
 * ```ts
 * const instructions = await changeTwitterRegistryData("bonfida", owner, 0, Buffer.from("data"));
 * ```
 */
export async function changeTwitterRegistryData(
  twitterHandle: string,
  verifiedPubkey: PublicKey,
  offset: number, // The offset at which to write the input data into the NameRegistryData
  input_data: Buffer,
): Promise<TransactionInstruction[]> {
  const hashedTwitterHandle = getHashedNameSync(twitterHandle);
  const twitterHandleRegistryKey = getNameAccountKeySync(
    hashedTwitterHandle,
    undefined,
    TWITTER_ROOT_PARENT_REGISTRY_KEY,
  );

  const instructions = [
    updateInstruction(
      NAME_PROGRAM_ID,
      twitterHandleRegistryKey,
      new Numberu32(offset),
      input_data,
      verifiedPubkey,
    ),
  ];

  return instructions;
}
